[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ciri/iese-dsfb/blob/main/notebooks/210-Amazon-Part-II.ipynb)

# Tabular: Amazon Part II

Last time, we identified a decline in revenue and overdependence on one category. Today, we'll investigate what might be hurting revenue (like cancellations), and then look at where and to whom Amazon could sell more — ending with a focused growth strategy.

In [1]:
import pandas as pd
df = pd.read_csv('../resources/tabular/Amazon-cleaned.csv')

## Cancellations & Returns

Not all orders are good news. In e-commerce, cancellations and returns create serious operational and financial issues:
* Lost revenue: These orders don’t convert into sales.
* Shipping and handling costs: Often non-refundable.
* Reverse logistics: Costly and time-consuming.
* Chargebacks: Especially if customers dispute transactions.
* Fraud & abuse: Repeated returns or B2C-to-B2B loopholes.
* Money laundering flags: In some cases, abnormal return behavior is a red flag.

Long story short: cancellations and returns hurt revenue.

Let’s measure the size of the problem:

* How many orders were cancelled?
* How many were returned?
* What percentage of total orders does this represent?

So far, you’ve used `.groupby()` for totals and averages. Now let’s introduce `.value_counts()` which is a quick way to count categories in one column, like a faster `.groupby('status').shape`.

In [2]:
# Step 1: Count all status values
# ---> same as: df.groupby('status').size().sort_values(ascending=False)
df['status'].value_counts() 

status
Shipped                          77593
Shipped - Delivered to Buyer     28761
Cancelled                        10766
Shipped - Returned to Seller      1950
Shipped - Picked Up                973
Pending                            656
Pending - Waiting for Pick Up      281
Shipped - Returning to Seller      145
Shipped - Out for Delivery          35
Shipped - Rejected by Buyer         11
Shipped - Lost in Transit            5
Shipped - Damaged                    1
Name: count, dtype: int64

In [3]:
# Step 2: Filter for problematic orders (cancelled or returned)
cancelled = df[df['status'].str.contains('Cancelled', case=False)]
returned  = df[df['status'].str.contains('Returned', case=False)]

In [4]:
# Step 3: Calculate totals and percentages
total_orders    = df.shape[0]
cancelled_pct   = cancelled.shape[0] / total_orders * 100
returned_pct    = returned.shape[0] / total_orders * 100
problematic_pct = (cancelled.shape[0]  + returned.shape[0]) / total_orders * 100

print(f"Cancelled Orders: {len(cancelled)} ({cancelled_pct:.2f}%)")
print(f"Returned Orders: {len(returned)} ({returned_pct:.2f}%)")
print(f"Total Problematic Orders: {problematic_pct:.2f}%")

Cancelled Orders: 10766 (8.88%)
Returned Orders: 1950 (1.61%)
Total Problematic Orders: 10.49%


> Observations: High cancellation rates (churn?), high return rates (issues with quality or fit? need to figure this out to avoid expensive logistics).

### Revenue by Customer Type (B2B vs. Regular)

Amazon doesn’t just sell to everyday consumers — it also serves **business buyers** through **Amazon Business**.

| **Customer Type** | **Examples**                                      | **What They Value**                                |
|--------------------|---------------------------------------------------|----------------------------------------------------|
| Regular (B2C)       | Individual shoppers buying for personal use      | Price, fast shipping, product variety              |
| Business (B2B)      | SMEs, schools, retailers buying in bulk          | Discounts, tax invoices, delivery reliability, APIs|

Business customers usually spend more per order — but place fewer orders.  
Regular customers are more frequent, but each transaction is smaller.

Let’s compare:
1. Total revenue by customer type  
2. Average order value (AOV) by customer type  
3. Number of orders by customer type

In [5]:
# Step 1: Group by customer type and summarize metrics
customer_summary = df.groupby('b2b').agg(
    total_revenue   = ('amount_eu', 'sum'),
    avg_order_value = ('amount_eu', 'mean'),
    order_count     = ('order_id', 'count')
).reset_index()
customer_summary

b2b  total_revenue  avg_order_value  order_count
0  False      857998.83         7.130145       120334
1   True        6503.49         7.714698          843

In [6]:
# Step 2: Convert True/False to readable labels
customer_summary['b2b'] = customer_summary['b2b'].replace({True: 'B2B', False: 'B2C'})
customer_summary


b2b  total_revenue  avg_order_value  order_count
0  B2C      857998.83         7.130145       120334
1  B2B        6503.49         7.714698          843

**Observations**:

| **Metric**         | **What It Shows**                                                                                |
|--------------------|--------------------------------------------------------------------------------------------------|
| **Order Count** | Regular customers dominate volume — this is the bulk of the business                            |
| **AOV**         | Business customers spend **~8% more per order** (€7.71 vs €7.13) but much less than expected.                                  |
| **Total Revenue** | Business is currently a *tiny slice* of total revenue (~0.75%) — but with higher ticket size |

Potential actionable insight: one last attempt at invigorating and attracting better B2B through e.g. exclusive deals, bulk pricing, repeate order flows? If it doesn't work we may consider spinning off or integrating it into B2C.

## Most Popular Products by State

Let's say that you’re working as a data scientist or BI developer. The goal is not just to analyze but rather to communicate insights clearly as we've been trying to do before. Sometimes a plot is enough, other times a dynamic dashboard is needed.

Dashboards allow business users (e.g., regional managers) to quickly:

* See how their region is doing
* Spot product trends
* Act on poor performance
* ..

We'll build a simple command-line dashboard first, and then make it interactive.

### More data cleaning!

In real data, categories like “state” or “country” often contain typos, inconsistent capitalization, or alternative names. Before we analyze sales by state, we need to standardize the state names so we don’t split data across 5 versions of “Rajasthan”, etc.

The code below we will skim over for now, but in the last part of the course we will explain how to handle text/categorical data better!

In [7]:
df.ship_state.unique()

array(['MAHARASHTRA', 'KARNATAKA', 'PUDUCHERRY', 'TAMIL NADU',
       'UTTAR PRADESH', 'CHANDIGARH', 'TELANGANA', 'ANDHRA PRADESH',
       'RAJASTHAN', 'DELHI', 'HARYANA', 'ASSAM', 'JHARKHAND',
       'CHHATTISGARH', 'ODISHA', 'KERALA', 'MADHYA PRADESH',
       'WEST BENGAL', 'NAGALAND', 'Gujarat', 'UTTARAKHAND', 'BIHAR',
       'JAMMU & KASHMIR', 'PUNJAB', 'HIMACHAL PRADESH',
       'ARUNACHAL PRADESH', 'MANIPUR', 'Goa', 'MEGHALAYA', 'GOA',
       'TRIPURA', 'LADAKH', 'DADRA AND NAGAR', 'SIKKIM', 'Delhi', nan,
       'ANDAMAN & NICOBAR ', 'Punjab', 'Rajshthan', 'Manipur',
       'rajasthan', 'Odisha', 'NL', 'Bihar', 'MIZORAM', 'punjab',
       'New Delhi', 'Rajasthan', 'Punjab/Mohali/Zirakpur', 'Puducherry',
       'delhi', 'RJ', 'Chandigarh', 'orissa', 'LAKSHADWEEP', 'goa', 'PB',
       'Arunachal Pradesh', 'AR', 'Pondicherry', 'Sikkim',
       'Arunachal pradesh', 'Nagaland', 'bihar', 'Mizoram', 'rajsthan',
       'Orissa', 'Rajsthan', 'Meghalaya'], dtype=object)

In [8]:
# clean up
# Mapping of messy names to cleaned names
df  = df.dropna()
df['ship_state'] = df['ship_state'].str.upper().str.strip()
state_map = {
    'DELHI': 'DELHI', 'NEW DELHI': 'DELHI', 'DELHI ': 'DELHI', 'DELHI/NCR': 'DELHI',
    'RAJASTHAN': 'RAJASTHAN', 'RAJSTHAN': 'RAJASTHAN', 'RAJSHAN': 'RAJASTHAN', 'RJ': 'RAJASTHAN',
    'ODISHA': 'ODISHA', 'ORISSA': 'ODISHA', 'ORISSA ': 'ODISHA',
    'PUNJAB/MOHALI/ZIRAKPUR': 'PUNJAB', 'PB': 'PUNJAB',
    'GOA': 'GOA', 'GOA ': 'GOA',
    'PUDUCHERRY': 'PUDUCHERRY', 'PONDICHERRY': 'PUDUCHERRY',
    'CHANDIGARH': 'CHANDIGARH',
    'ARUNACHAL PRADESH': 'ARUNACHAL PRADESH', 'ARUNACHAL PRADESH ': 'ARUNACHAL PRADESH', 
    'AR': 'ARUNACHAL PRADESH',
    'SIKKIM': 'SIKKIM',
    'MEGHALAYA': 'MEGHALAYA',
    'NAGALAND': 'NAGALAND', 'NL': 'NAGALAND',
    'MIZORAM': 'MIZORAM',
    'BIHAR': 'BIHAR', 'BIHAR ': 'BIHAR', 'Bihar': 'BIHAR', 'bihar': 'BIHAR',
}

df['ship_state'] = df['ship_state'].replace(state_map)
df['ship_state'].unique()

array(['KARNATAKA', 'TELANGANA', 'MAHARASHTRA', 'ANDHRA PRADESH',
       'HARYANA', 'JHARKHAND', 'CHHATTISGARH', 'ODISHA', 'UTTAR PRADESH',
       'TAMIL NADU', 'WEST BENGAL', 'RAJASTHAN', 'MADHYA PRADESH',
       'ASSAM', 'KERALA', 'DELHI', 'BIHAR', 'UTTARAKHAND', 'MEGHALAYA',
       'PUNJAB', 'GUJARAT', 'JAMMU & KASHMIR', 'DADRA AND NAGAR',
       'PUDUCHERRY', 'TRIPURA', 'CHANDIGARH', 'GOA', 'ANDAMAN & NICOBAR',
       'HIMACHAL PRADESH', 'MANIPUR', 'SIKKIM', 'MIZORAM', 'NAGALAND',
       'ARUNACHAL PRADESH', 'LADAKH', 'LAKSHADWEEP'], dtype=object)

Let's focus on one state to understand its sales profile. We'll start with Karnataka, one of the largest markets

In [9]:
# Filter data for Karnataka
karnataka = df[df['ship_state'] == 'KARNATAKA']

In [10]:
# Calculate the number of orders placed in this state
total_orders = len(karnataka)

# Calculate total revenue generated
total_revenue = karnataka['amount_eu'].sum()

# Average revenue per order — helps identify spending behavior
avg_order_value = karnataka['amount_eu'].mean()

# Find the most popular product category by quantity sold
top_category = (
    karnataka.groupby('category')['qty']
    .sum()
    .sort_values(ascending=False)
    .head(1)
)

# Print the summary stats
print(f"Total Orders: {total_orders}")
print(f"Total Revenue: €{total_revenue:,.2f}")
print(f"Average Order Value: €{avg_order_value:,.2f}")
print(f"Top Category: {top_category.index[0]} ({top_category.values[0]} units)")

Total Orders: 4001
Total Revenue: €27,698.96
Average Order Value: €6.92
Top Category: kurta (1527 units)


Now, we wrap whatever you came up with in a function

In [11]:
def draw_state_dashboard(state):
    # Standardize input casing
    subset = df[df['ship_state'] == state.upper()]
    
    if subset.empty:
        print("No data for selected state.")
        return

    # Core KPIs
    total_orders = len(subset)
    total_revenue = subset['amount_eu'].sum()
    avg_order_value = subset['amount_eu'].mean()
    top_category = (
        subset.groupby('category')['qty']
        .sum()
        .sort_values(ascending=False)
        .head(1)
    )

    # Print the summary
    print(f"--- Dashboard for {state.title()} ---")
    print(f"Total Orders: {total_orders}")
    print(f"Total Revenue: €{total_revenue:,.2f}")
    print(f"Average Order Value: €{avg_order_value:,.2f}")
    print(f"Top Category: {top_category.index[0]} ({top_category.values[0]} units)")

draw_state_dashboard("karnataka")

--- Dashboard for Karnataka ---
Total Orders: 4001
Total Revenue: €27,698.96
Average Order Value: €6.92
Top Category: kurta (1527 units)


**Dashboard**

Wonderful, all that is needed at this point is to make it interactive with a dropdown. Note: this code assumes that you are using a notebook like google collab or jupyter lab. Notably it won't work in VS Code.

Note: You will need to install and enable widgets, it doesn't always work on every environment.

In [12]:
! pip install ipywidgets==8.0.2
! pip install jupyterlab_widgets==3.0.13

  Attempting uninstall: ipywidgets
    Found existing installation: ipywidgets 8.1.5
    Uninstalling ipywidgets-8.1.5:
      Successfully uninstalled ipywidgets-8.1.5


In [26]:
import ipywidgets as widgets
from IPython.display import display, clear_output

# Dropdown with all cleaned states
output = widgets.Output()
state_dropdown = widgets.Dropdown(
    options=sorted(df['ship_state'].dropna().unique()),
    description='Select a State:',
    style={'description_width': 'initial'}
)

# When a user selects a different state, redraw the dashboard
def on_state_change(change):
    if change['type'] == 'change' and change['name'] == 'value':
       with output:
            clear_output(wait=True)
            draw_state_dashboard(change['new'])

# Connect the dropdown to the callback
state_dropdown.observe(on_state_change)
display(state_dropdown, output)

Dropdown(description='Select a State:', options=('ANDAMAN & NICOBAR', 'ANDHRA PRADESH', 'ARUNACHAL PRADESH', '…

Output()

### BONUS: add some analysis from these and the previous classes to make it prettier

In [56]:
# Define enhanced dashboard function
df['date'] = pd.to_datetime(df['date'])

def draw_state_dashboard(state):
    subset = df[df['ship_state'] == state.upper()]
    if subset.empty:
        print("No data for selected state.")
        return

    # Set up 2x2 subplot layout
    fig, axs = plt.subplots(2, 2, figsize=(15, 10))

    # Pane 1: Bar plot of top categories
    top_cats = (subset.groupby('category')['qty']
                .sum()
                .sort_values(ascending=False)
                .head(6)
                .reset_index())
    sns.barplot(data=top_cats, x='category', y='qty', ax=axs[0, 0], palette='Blues_d', hue='category')
    axs[0, 0].set_title("Top Product Categories")
    axs[0, 0].set_ylabel("Quantity Sold")
    axs[0, 0].set_xlabel("")
    axs[0, 0].tick_params(axis='x', rotation=45)

    # Pane 2: Monthly revenue trend
    monthly = (subset.groupby(subset['date'].dt.to_period('M'))['amount_eu']
               .sum()
               .sort_index())
    monthly.index = monthly.index.to_timestamp()
    axs[0, 1].plot(monthly.index, monthly.values, marker='o')
    axs[0, 1].set_title("Monthly Revenue Trend")
    axs[0, 1].set_ylabel("Revenue")
    axs[0, 1].set_xlabel("Month")
    axs[0, 1].tick_params(axis='x', rotation=45)
    
    # Pane 3: Order status breakdown as a bar plot
    status_counts = (subset['status']
                     .value_counts()
                     .head(6)
                     .sort_values(ascending=False)
                     .reset_index())
    
    sns.barplot(data=status_counts, y='status', x='count', ax=axs[1, 0], palette='pastel', hue='status', orient='h')
    
    axs[1, 0].set_title("Top 5 Order Statuses")
    axs[1, 0].set_ylabel("Order Count")
    axs[1, 0].set_xlabel("")
    axs[1, 0].tick_params(axis='x', rotation=45)


    # Pane 4: Show summary stats as table
    total_orders = len(subset)
    total_revenue = subset['amount_eu'].sum()
    avg_order_value = subset['amount_eu'].mean()
    b2b_share = subset['b2b'].mean() * 100
    top_category = top_cats.iloc[0]['category']
    summary_data = pd.DataFrame({
        'Metric': ['Total Orders', 'Total Revenue', 'Average Order Value', 'B2B Share (%)', 'Top Category'],
        'Value': [total_orders, f"€{total_revenue:,.0f}", f"€{avg_order_value:,.0f}", f"{b2b_share:.1f}%", top_category]
    })
    axs[1, 1].axis('off')
    table = axs[1, 1].table(cellText=summary_data.values,
                           colLabels=summary_data.columns,
                           cellLoc='left', loc='center')
    table.auto_set_font_size(False)
    table.set_fontsize(10)
    table.scale(1, 1.5)

    plt.tight_layout()
    plt.show()

# UI Dropdown
output = widgets.Output()
state_dropdown = widgets.Dropdown(
    options=sorted(df['ship_state'].dropna().unique()),
    description='Select a State:',
    style={'description_width': 'initial'}
)

def on_state_change(change):
    if change['type'] == 'change' and change['name'] == 'value':
        with output:
            clear_output(wait=True)
            draw_state_dashboard(change['new'])

state_dropdown.observe(on_state_change)
display(state_dropdown, output)

Dropdown(description='Select a State:', options=('ANDAMAN & NICOBAR', 'ANDHRA PRADESH', 'ARUNACHAL PRADESH', '…

Output()